ot/smooth.py

```diff
@@ -636,3 +683,166 @@ def smooth_ot_semi_dual(a, b, M, reg, reg_type='l2', method="L-BFGS-B", stopThr=
         return G, log
     else:
         return G
+
+
+def sparsity_constrained_ot_dual(
```

Comment:
You could reuse the existing functions with `reg_type="sparse_l2"` or `reg_type="l2_topk"` instead. You would need to add `max_nz` but I think that's fine.